### PR TITLE
[openssh] Introduce custom openssh-server package for supporting reverse console SSH

### PR DESCRIPTION
**Pull request by Blueve**
**- Why I did it**

To support reverse SSH feature, we need to build a custom openssh-server package and replace the official one in SONiC OS.

(see HLD for more detail: https://github.com/Azure/SONiC/blob/master/doc/console/SONiC-Console-Switch-High-Level-Design.md#313-reverse-ssh)

**- How I did it**

- Modify source code of openssh-server to allow the sshd expose the `style` field in `$SSH_TARGET_CONSOLE_LINE`.
- Build custom openssh-server
- Install custom openssh-server

**- How to verify it**

- [x] Build custom openssh-server package and install/test it on a physical SONiC switch
- [x] Build an image which includes the change of current PR and install/test it on a physical DUT

Test Steps:
1. Login with below options
image

1. Run below command and check the result
```bash
admin@sonic:~$ echo $SSH_TARGET_CONSOLE_LINE
1
```

**- Which release branch to backport (provide reason below if selected)**

- [ ] 201811
- [ ] 201911
- [ ] 202006

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch [hunk 1]
```diff
@@ -0,0 +1,2651 @@
+From e0f1ac19ec241a54418dbe8e5eaf4a7595b62287 Mon Sep 17 00:00:00 2001
+From: Jing Kan <jika@microsoft.com>
+Date: Wed, 21 Oct 2020 20:14:41 +0800
+Subject: [PATCH] Put style as line number to ssh session environment variable
+
+---
+ config.guess | 586 +++++++++++++++++++++++++++------------------------
```
Comment (lguohan) [inline src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch:7]: the patch contains lots of unrelated changes. please simplify.

**Review by lguohan — CHANGES_REQUESTED** [review]
the patch contains lots of unrelated changes. please simplify.

File: src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch [hunk 2]
```diff
@@ -0,0 +1,2651 @@
...
+ 	    *)
+-		release=`echo ${UNAME_RELEASE} | sed -e 's/[-_].*//' | cut -d. -f1,2`
++		release=`echo "$UNAME_RELEASE" | sed -e 's/[-_].*//' | cut -d. -f1,2`
+ 		;;
+ 	esac
+ 	# Since CPU_TYPE-MANUFACTURER-KERNEL-OPERATING_SYSTEM:
+ 	# contains redundant information, the shorter form:
+ 	# CPU_TYPE-MANUFACTURER-OPERATING_SYSTEM is used.
+-	echo "${machine}-${os}${release}${abi}"
++	echo "$machine-${os}${release}${abi}"
```
Comment (lguohan) [inline src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch:189]: not sure why it is changed. lots of such example. please remove these unrelated changes and only keep the one is related to your needed function.

**Comment by lguohan** [comment]
i am not able to review the patch till you have simplfied the patch.

File: src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch (hunk 1 above)
Comment (Blueve) [reply src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch:7]: Yes... I will update it .

File: src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch (hunk 2 above)
Comment (Blueve) [reply src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch:189]: We should only include changes in 'session.c'. I will update the patch. Not sure why the commit includes those...

**Comment by Blueve** [comment]
Seems like the openssh-server package were not been built with image.

```
dpkg: error: cannot access archive 'target/debs/buster/openssh-server_*.deb': No such file or directory
```

Investigating...

But run below command works.
```
target/debs/buster/openssh-server_7.9p1-10+deb10u2_amd64.deb
```

**Comment by Blueve** [comment]
retest this please

**Comment by Blueve** [comment]
retest this please

File: src/openssh/Makefile [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+MAIN_TARGET = openssh-server_$(OPENSSH_VERSION)_$(CONFIGURED_ARCH).deb
+DERIVED_TARGETS = openssh-server-dbgsym_$(OPENSSH_VERSION)_$(CONFIGURED_ARCH).deb
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% :
+	# Obtain openssh: https://salsa.debian.org/ssh-team/openssh/-/tree/debian/1%257.9p1-10+deb10u2
+	rm -rf ./openssh-server
+	git clone https://salsa.debian.org/ssh-team/openssh.git openssh-server
+	pushd ./openssh-server
+	git checkout -b openssh-src -f 6d9ca74c48d9911342c6ca5aaac8a25974fa2619
```
Comment (lguohan) [inline src/openssh/Makefile:13]: does openssh have a tag for this commit id?

Comment (Blueve) [reply src/openssh/Makefile:13]: Yes, they have. The commit id is targeting to tag: `debian/1%257.9p1-10+deb10u2`

Comment (lguohan) [reply src/openssh/Makefile:13]: can you use the tag id or add a comment.

File: src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch [hunk 4]
```diff
@@ -0,0 +1,28 @@
+From 839f98d579e3a22b6ca075fd026eee55bdb2ef25 Mon Sep 17 00:00:00 2001
+From: Blueve <672454911@qq.com>
+Date: Mon, 26 Oct 2020 06:44:59 +0000
+Subject: [PATCH] Put style as line number to ssh session environment variable
+
```
Comment (lguohan) [inline src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch:5]: can you put following description in this patch. this is from your design doc.

```
By default, the content between : and @ will be trimmed by sshd before it do authentication and the trimmed string will be dropped silently. To use this segment as line number for reverse SSH feature, we need to modify the source ode of OpenSSH and put this segment to a environment variable SSH_TARGET_CONSOLE_LINE, then we can insert a short script into /etc/bash.bashrc and run command consutil connect $SSH_TARGET_CONSOLE_LINE to enter the management session automatically after user login.
```

**Comment by lguohan** [comment]
@qiluo-msft, please take a look at this openssh change.

**Review by lguohan — APPROVED** [review]

File: src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch (hunk 4 above)
Comment (Blueve) [reply src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch:5]: Sure, I will update them

File: src/openssh/Makefile (hunk 3 above)
Comment (Blueve) [reply src/openssh/Makefile:13]: Yes, I will update

Comment (yxieca) [reply src/openssh/Makefile:13]: @Blueve I didn't see the the change to checkout tag yet. did you forget to push?

File: src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch [hunk 5]
```diff
@@ -0,0 +1,28 @@
+From 839f98d579e3a22b6ca075fd026eee55bdb2ef25 Mon Sep 17 00:00:00 2001
+From: Blueve <672454911@qq.com>
```
Comment (qiluo-msft) [inline src/openssh/patch/0001-Put-style-as-line-number-to-ssh-session-environment-.patch:2]: Suggest use
```
From: Blueve <blueve@users.noreply.github.com>
```

File: rules/openssh.dep [hunk 6]
```diff
@@ -0,0 +1,10 @@
+
+SPATH       := $($(OPENSSH_SERVER)_SRC_PATH)
```
Comment (qiluo-msft) [inline rules/openssh.dep:2]: Remove the empty lines at the header and the end of the file.

**Review by qiluo-msft — COMMENTED** [review]
As comments

File: src/openssh/Makefile (hunk 3 above)
Comment (Blueve) [reply src/openssh/Makefile:13]: I didn't have time yesterday. I will update this PR today.

**Review by yxieca — DISMISSED** [review]

File: src/openssh/Makefile [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+MAIN_TARGET = openssh-server_$(OPENSSH_VERSION)_$(CONFIGURED_ARCH).deb
+DERIVED_TARGETS = openssh-server-dbgsym_$(OPENSSH_VERSION)_$(CONFIGURED_ARCH).deb
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% :
+	# Obtain openssh: https://salsa.debian.org/ssh-team/openssh/-/tree/debian/1%257.9p1-10+deb10u2
+	rm -rf ./openssh-server
+	git clone https://salsa.debian.org/ssh-team/openssh.git openssh-server
+	pushd ./openssh-server
+	git checkout -b openssh-src -f debian/1\%$(OPENSSH_VERSION)
```
Comment (qiluo-msft) [inline src/openssh/Makefile:13]: >debian/1\%$(OPENSSH_VERSION) [](start = 32, length = 28)

The commit ID is better than tag because it is not movable and reproducible.
@xumia Please help check.

**Review by qiluo-msft — CHANGES_REQUESTED** [review]
As comment

File: src/openssh/Makefile (hunk 7 above)
Comment (Blueve) [reply src/openssh/Makefile:13]: Talk with Xuhui offline. I will use a commit id here for better reproducible and add a comment to specific the tag name

**Review by qiluo-msft — APPROVED** [review]

**Comment by Blueve** [comment]
retest vsimage please

**Comment by Blueve** [comment]
retest vsimage please

**Comment by Blueve** [comment]
retest vsimage please

**Comment by Blueve** [comment]
retest vsimage please
